UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,905 @@
+  IN VOID   *Fdt,
+  IN INT32  Node,
+  IN UINT8  PciEnumDone,
+  IN UINT8  RootBridgeCount,
+  IN CHAR8  *GmaStr,
+  IN UINT8  *index
+  )
+{
+  INT32               SubNode;
+  INT32               Property;
+  INT32               SSubNode;
+  FDT_NODE_HEADER     *NodePtr;
+  CONST FDT_PROPERTY  *PropertyPtr;
+  INT32               TempLen;
+  UINT32              *Data32;
+  UINT32              MemType;
+  CONST CHAR8         *TempStr;
+  UINT8               RbIndex;
+  UINTN               HobDataSize;
+  UINT8               Base;
+
+  RbIndex     = *index;
+  HobDataSize = sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES) + RootBridgeCount *sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGE);
+  //
+  // Create PCI Root Bridge Info Hob.
+  //
+  if (mPciRootBridgeInfo == NULL) {
+    mPciRootBridgeInfo = BuildGuidHob (&gUniversalPayloadPciRootBridgeInfoGuid, HobDataSize);
+    ASSERT (mPciRootBridgeInfo != NULL);
+    if (mPciRootBridgeInfo == NULL) {
+      return;
+    }
+
+    ZeroMem (mPciRootBridgeInfo, HobDataSize);
+    mPciRootBridgeInfo->Header.Length    = sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES);
+    mPciRootBridgeInfo->Header.Revision  = UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES_REVISION;
+    mPciRootBridgeInfo->Count            = RootBridgeCount;
+    mPciRootBridgeInfo->ResourceAssigned = (BOOLEAN)PciEnumDone;
+  }
+
+  for (SubNode = FdtFirstSubnode (Fdt, Node); SubNode >= 0; SubNode = FdtNextSubnode (Fdt, SubNode)) {
+    NodePtr = (FDT_NODE_HEADER *)((CONST CHAR8 *)Fdt + SubNode + Fdt32ToCpu (((FDT_HEADER *)Fdt)->OffsetDtStruct));
+    DEBUG ((DEBUG_INFO, "\n      SubNode(%08X)  %a", SubNode, NodePtr->Name));
+
+    if (AsciiStrnCmp (NodePtr->Name, GmaStr, AsciiStrLen (GmaStr)) == 0) {
+      DEBUG ((DEBUG_INFO, "  Found gma@ node \n"));
+      ParsegraphicNode (Fdt, SubNode);
+    }
+
+    if (AsciiStrnCmp (NodePtr->Name, "isa", AsciiStrLen ("isa")) == 0) {
+      SSubNode = FdtFirstSubnode (Fdt, SubNode); // serial
+      ParseSerialPort (Fdt, SSubNode);
+    }
+
+    if (AsciiStrnCmp (NodePtr->Name, "serial@", AsciiStrLen ("serial@")) == 0) {
+      ParseSerialPort (Fdt, SubNode);
+    }
+  }
+
+  DEBUG ((DEBUG_INFO, " RbIndex :%x \n", RbIndex));
+
+  for (Property = FdtFirstPropertyOffset (Fdt, Node); Property >= 0; Property = FdtNextPropertyOffset (Fdt, Property)) {
+    PropertyPtr = FdtGetPropertyByOffset (Fdt, Property, &TempLen);
+    TempStr     = FdtGetString (Fdt, Fdt32ToCpu (PropertyPtr->NameOffset), NULL);
+
+    if (AsciiStrCmp (TempStr, "ranges") == 0) {
+      DEBUG ((DEBUG_INFO, "  Found ranges Property TempLen (%08X), limit %x\n", TempLen, TempLen/sizeof (UINT32)));
+
+      mPciRootBridgeInfo->RootBridge[RbIndex].AllocationAttributes = EFI_PCI_HOST_BRIDGE_COMBINE_MEM_PMEM | EFI_PCI_HOST_BRIDGE_MEM64_DECODE;
+      mPciRootBridgeInfo->RootBridge[RbIndex].Supports             = ROOT_BRIDGE_SUPPORTS_DEFAULT;
+      mPciRootBridgeInfo->RootBridge[RbIndex].PMemAbove4G.Base     = PcdGet64 (PcdPciReservedPMemAbove4GBBase);

```

Comment:
I realized DmaAbove4G is another important param. We need to pass it from DT or PCD?